language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToFunctionCodeAction.java

```diff
@@ -0,0 +1,593 @@
+                .filter(symbol -> symbol.kind() == SymbolKind.VARIABLE || symbol.kind() == SymbolKind.PARAMETER)
+                .filter(symbol -> context.currentSemanticModel().get().references(symbol).stream()
+                        .anyMatch(location -> PositionUtil.isRangeWithinRange(PositionUtil
+                                        .getRangeFromLineRange(location.lineRange()),
+                                PositionUtil.toRange(matchedLineRange))))
+                // for every variable and parameter symbol, the symbol.getLocation() is not empty
+                .filter(symbol -> !PositionUtil.isWithinLineRange(symbol.getLocation().get().lineRange(),
+                        matchedLineRange))
+                .sorted(Comparator.comparingInt(symbol -> symbol.getLocation().get().textRange().startOffset()))
+                .collect(Collectors.toList());
+    }
+
+    private Optional<ArgLists> getArgLists(CodeActionContext context, List<Symbol> symbolsList) {
+        List<String> argsForExtractFunction = new ArrayList<>();
+        List<String> argsForReplaceFunctionCall = new ArrayList<>();
+
+        for (Symbol symbol : symbolsList) {
+            if (symbol.getName().isEmpty()) {
+                return Optional.empty();
+            }
+            argsForReplaceFunctionCall.add(symbol.getName().get());
+
+            if (symbol.kind() == SymbolKind.VARIABLE) {
+                Optional<String> possibleType = CodeActionUtil
+                        .getPossibleType(((VariableSymbol) symbol).typeDescriptor(), new ArrayList<>(), context);
+                if (possibleType.isEmpty()) {
+                    return Optional.empty();
+                }
+                argsForExtractFunction.add(String.format("%s %s", possibleType.get(), symbol.getName().get()));
+            } else if (symbol.kind() == SymbolKind.PARAMETER) {
+                Optional<String> possibleType = CodeActionUtil
+                        .getPossibleType(((ParameterSymbol) symbol).typeDescriptor(), new ArrayList<>(), context);
+                if (possibleType.isEmpty()) {
+                    return Optional.empty();
+                }
+                argsForExtractFunction.add(String.format("%s %s", possibleType.get(), symbol.getName().get()));
+            }
+        }
+
+        ArgLists argLists = new ArgLists();
+        argLists.extractFunctionArgs = argsForExtractFunction;
+        argLists.replaceFunctionCallArgs = argsForReplaceFunctionCall;
+
+        return Optional.of(argLists);
+    }
+
+    private Optional<Node> findEnclosingModulePartNode(NonTerminalNode node) {
+        Node reference = node;
+
+        while (reference != null && reference.parent() != null) {
+            if (reference.parent().kind() == SyntaxKind.MODULE_PART) {
+                return Optional.of(reference);
+            }
+            reference = reference.parent();
+        }
+
+        return Optional.empty();
```

Comment:
@TharushiJay  Can you please share how that can be done efficiently?

Reply:
```
while (reference != null && reference.parent() != null
                && reference.parent().kind() != SyntaxKind.MODULE_PART) {
            reference = reference.parent();
        }
        return Optional.ofNullable(reference);
```

Reply:
Yeah so we cannot return a `Node` right?

Reply:
If we use your logic to this function, we can get returned,

1. The root node
2. A module level node

The whole idea of this function is to find the module level node if there's one enclosing the node, or else return empty. So with this logic(), I need to check whether the returned node(assuming returning reference instead of ofNullable(reference)) is a ModulePart or not. In that case, the function does not serve the objective. IMHO we should stick to this implementation where I don't even see a significant performance difference.

Reply:
Seems like we can simplify this further.
```
while (reference.parent() != null && reference.parent().kind() != SyntaxKind.MODULE_PART) {
            reference = reference.parent();
        }
return reference;
```
Removed the null check for `reference`, since at the first iteration `matchedCodeActionNode` won't be null (already validated) and for the subsequent iterations `reference.parent() != null` handles it, right?
Either way the loop will only exit when it's null or if it has found a node that satisfies the required condition. Instead of checking for empty, having a null check where this method is called would work.
As far as I can see this returns the same thing that your implementation does. A null and a reference instead of empty and a reference. Can you give me an example if not? Not sure about the performance but it simplifies the logic and improves readability.

Reply:
Yeah the root node cannot be a matched code action since your point is correct +1